service/matching/version_rule_helpers.go

```diff
@@ -491,6 +501,51 @@ func dfs(curr string, visited, inStack map[string]bool, nodes map[string][]strin
 	return false
 }
 
+// getUpstreamBuildIds returns a list of build ids that point to the given buildId in the graph of redirect rules.
+// It considers all rules in the input list, so the caller should provide a filtered list as needed.
+// The result will contain no duplicates.
+func getUpstreamBuildIds(buildId string, redirectRules []*persistencespb.RedirectRule) []string {
+	return getUpstreamHelper(buildId, redirectRules, nil)
+}
+
+func getUpstreamHelper(
+	buildId string,
+	redirectRules []*persistencespb.RedirectRule,
+	visited []string) []string {
+	var upstream []string
+	visited = append(visited, buildId)
+	directSources := getSourcesForTarget(buildId, redirectRules)
+
+	for _, src := range directSources {
+		if !slices.Contains(visited, src) {
+			upstream = append(upstream, src)
+			upstream = append(upstream, getUpstreamHelper(src, redirectRules, visited)...)
+		}
+	}
```

Comment:
This is at least `O(n^2)` (ignoring the recursion for a moment). At a bare minimum you could have `visited` be a map so you could replace the `O(n)` slices.Contains call with an O(1) map lookup

Reply:
we enforce that the maximum chain length of the graph is 50, which ensures that `visited` will never get that big, so I didn't consider it a priority to optimize this. But I can definitely improve it

Reply:
I swapped it to a map